From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ishland <30385023+ishland@users.noreply.github.com>
Date: Fri, 10 Jul 2026 01:07:43 +0800
Subject: [PATCH] Port 26.1.2 -> 1.21.11


diff --git a/build.gradle b/build.gradle
index c680a91..1caff61 100644
--- a/build.gradle
+++ b/build.gradle
@@ -1,5 +1,5 @@
 plugins {
-    id 'net.fabricmc.fabric-loom' version '1.17-SNAPSHOT'
+    id 'net.fabricmc.fabric-loom-remap' version '1.17-SNAPSHOT'
     id 'maven-publish'
     id 'com.gradleup.shadow' version '9.4.0'
     id 'me.modmuss50.mod-publish-plugin' version "2.1.1"
@@ -57,7 +57,8 @@ base {
 dependencies {
     //to change the versions see the gradle.properties file
     minecraft "com.mojang:minecraft:${project.minecraft_version}"
-    implementation "net.fabricmc:fabric-loader:${project.loader_version}"
+    mappings loom.officialMojangMappings()
+    modImplementation "net.fabricmc:fabric-loader:${project.loader_version}"
 
     // Fabric API. This is technically optional, but you probably want it anyway.
     //modImplementation "net.fabricmc.fabric-api:fabric-api:${project.fabric_version}"
@@ -100,8 +101,8 @@ java {
     // If you remove this line, sources will not be generated.
     withSourcesJar()
 
-    sourceCompatibility = JavaVersion.VERSION_25
-    targetCompatibility = JavaVersion.VERSION_25
+    sourceCompatibility = JavaVersion.VERSION_21
+    targetCompatibility = JavaVersion.VERSION_21
 }
 jar {
     archiveClassifier = "preshadow"
@@ -109,12 +110,19 @@ jar {
 }
 
 shadowJar {
-    archiveClassifier = "all"
+    archiveClassifier = "all-dev"
     configurations = [ project.configurations.shadowInclude ]
 
     relocate "com.ishland.flowsched", "ca.spottedleaf.starlight.interndep.flowsched"
 }
 
+remapJar {
+    input = shadowJar.archiveFile
+    archiveFileName = shadowJar.archiveFileName.get().replaceAll("-dev\\.jar\$", ".jar")
+    addNestedDependencies = true
+    dependsOn shadowJar
+}
+
 clean.dependsOn gradle.includedBuild('FlowSched').task(':clean')
 
 // make build reproducible
@@ -180,7 +188,7 @@ publishMods {
 
         curseforge("curseforgePackage") {
             from cfOptions
-            file = shadowJar.archiveFile
+            file = remapJar.archiveFile
             projectId = "1055925"
         }
     }
@@ -193,7 +201,7 @@ publishMods {
 
         modrinth("modrinthPackage") {
             from mrOptions
-            file = shadowJar.archiveFile
+            file = remapJar.archiveFile
             projectId = "Ps1zyz6x"
         }
     }
diff --git a/gradle.properties b/gradle.properties
index 89bc5f6..d3ff564 100644
--- a/gradle.properties
+++ b/gradle.properties
@@ -3,11 +3,11 @@ org.gradle.jvmargs=-Xmx2G
 org.gradle.daemon=false
 # Fabric Properties
 # check these on https://modmuss50.me/fabric.html
-minecraft_version=26.1.2
+minecraft_version=1.21.11
 loader_version=0.19.3
 # Mod Properties
 mod_version=0.3.0-alpha.0
 maven_group=ca.spottedleaf.starlight
 archives_base_name=ScalableLux-fabric
 
-curseforge_minecraft_versions=26.1.2
+curseforge_minecraft_versions=1.21.11
diff --git a/src/main/java/ca/spottedleaf/starlight/common/light/BlockStarLightEngine.java b/src/main/java/ca/spottedleaf/starlight/common/light/BlockStarLightEngine.java
index 8ea57dc..4d9e0ec 100644
--- a/src/main/java/ca/spottedleaf/starlight/common/light/BlockStarLightEngine.java
+++ b/src/main/java/ca/spottedleaf/starlight/common/light/BlockStarLightEngine.java
@@ -133,7 +133,7 @@ public final class BlockStarLightEngine extends StarLightEngine {
 
         final int sectionOffset = this.chunkSectionIndexOffset;
         final BlockState conditionallyOpaqueState;
-        int opacity = Math.max(1, centerState.getLightDampening());
+        int opacity = Math.max(1, centerState.getLightBlock());
         if (opacity >= 15) {
             return level; // TODO in older starlight, opacity >= 15 can go into the loop below, but it probably shouldn't
         }
@@ -195,8 +195,8 @@ public final class BlockStarLightEngine extends StarLightEngine {
     protected List<BlockPos> getSources(final LightChunkGetter lightAccess, final ChunkAccess chunk) {
         final List<BlockPos> sources = new ArrayList<>();
 
-        final int offX = chunk.getPos().x() << 4;
-        final int offZ = chunk.getPos().z() << 4;
+        final int offX = chunk.getPos().x << 4;
+        final int offZ = chunk.getPos().z << 4;
 
         final LevelChunkSection[] sections = chunk.getSections();
         for (int sectionY = this.minSection; sectionY <= this.maxSection; ++sectionY) {
diff --git a/src/main/java/ca/spottedleaf/starlight/common/light/SkyStarLightEngine.java b/src/main/java/ca/spottedleaf/starlight/common/light/SkyStarLightEngine.java
index 225752f..a391bc8 100644
--- a/src/main/java/ca/spottedleaf/starlight/common/light/SkyStarLightEngine.java
+++ b/src/main/java/ca/spottedleaf/starlight/common/light/SkyStarLightEngine.java
@@ -237,8 +237,8 @@ public final class SkyStarLightEngine extends StarLightEngine {
                                    final int toSection) {
         Arrays.fill(this.nullPropagationCheckCache, false);
         this.rewriteNibbleCacheForSkylight(chunk);
-        final int chunkX = chunk.getPos().x();
-        final int chunkZ = chunk.getPos().z();
+        final int chunkX = chunk.getPos().x;
+        final int chunkZ = chunk.getPos().z;
         for (int y = toSection; y >= fromSection; --y) {
             this.checkNullSection(chunkX, y, chunkZ, true);
         }
@@ -250,8 +250,8 @@ public final class SkyStarLightEngine extends StarLightEngine {
     protected void checkChunkEdges(final LightChunkGetter lightAccess, final ChunkAccess chunk, final ShortCollection sections) {
         Arrays.fill(this.nullPropagationCheckCache, false);
         this.rewriteNibbleCacheForSkylight(chunk);
-        final int chunkX = chunk.getPos().x();
-        final int chunkZ = chunk.getPos().z();
+        final int chunkX = chunk.getPos().x;
+        final int chunkZ = chunk.getPos().z;
         for (final ShortIterator iterator = sections.iterator(); iterator.hasNext();) {
             final int y = (int)iterator.nextShort();
             this.checkNullSection(chunkX, y, chunkZ, true);
@@ -305,7 +305,7 @@ public final class SkyStarLightEngine extends StarLightEngine {
 
         final BlockState conditionallyOpaqueState;
         this.recalcCenterPos.set(worldX, worldY, worldZ);
-        int opacity = Math.max(1, centerState.getLightDampening());
+        int opacity = Math.max(1, centerState.getLightBlock());
         if (((ExtendedAbstractBlockState)centerState).scalablelux$isConditionallyFullOpaque()) {
             conditionallyOpaqueState = centerState;
         } else {
@@ -359,8 +359,8 @@ public final class SkyStarLightEngine extends StarLightEngine {
         Arrays.fill(this.nullPropagationCheckCache, false);
 
         final BlockGetter world = lightAccess.getLevel();
-        final int chunkX = atChunk.getPos().x();
-        final int chunkZ = atChunk.getPos().z();
+        final int chunkX = atChunk.getPos().x;
+        final int chunkZ = atChunk.getPos().z;
         final int heightMapOffset = chunkX * -16 + (chunkZ * (-16 * 16));
 
         // setup heightmap for changes
@@ -454,8 +454,8 @@ public final class SkyStarLightEngine extends StarLightEngine {
 
         final BlockGetter world = lightAccess.getLevel();
         final ChunkPos chunkPos = chunk.getPos();
-        final int chunkX = chunkPos.x();
-        final int chunkZ = chunkPos.z();
+        final int chunkX = chunkPos.x;
+        final int chunkZ = chunkPos.z;
 
         final LevelChunkSection[] sections = chunk.getSections();
 
@@ -531,10 +531,10 @@ public final class SkyStarLightEngine extends StarLightEngine {
 
         if (highestNonEmptySection >= this.minSection) {
             // fill out our other sources
-            final int minX = chunkPos.x() << 4;
-            final int maxX = chunkPos.x() << 4 | 15;
-            final int minZ = chunkPos.z() << 4;
-            final int maxZ = chunkPos.z() << 4 | 15;
+            final int minX = chunkPos.x << 4;
+            final int maxX = chunkPos.x << 4 | 15;
+            final int minZ = chunkPos.z << 4;
+            final int maxZ = chunkPos.z << 4 | 15;
             final int startY = highestNonEmptySection << 4 | 15;
             for (int currZ = minZ; currZ <= maxZ; ++currZ) {
                 for (int currX = minX; currX <= maxX; ++currX) {
@@ -664,7 +664,7 @@ public final class SkyStarLightEngine extends StarLightEngine {
                     flags |= FLAG_HAS_SIDED_TRANSPARENT_BLOCKS;
                 }
 
-                final int opacity = current.getLightDampening();
+                final int opacity = current.getLightBlock();
                 if (opacity > 0) {
                     // let the queued value (if any) handle it from here.
                     break;
diff --git a/src/main/java/ca/spottedleaf/starlight/common/light/StarLightEngine.java b/src/main/java/ca/spottedleaf/starlight/common/light/StarLightEngine.java
index 3bfaceb..6785910 100644
--- a/src/main/java/ca/spottedleaf/starlight/common/light/StarLightEngine.java
+++ b/src/main/java/ca/spottedleaf/starlight/common/light/StarLightEngine.java
@@ -578,8 +578,8 @@ public abstract class StarLightEngine {
 
     protected void checkChunkEdges(final LightChunkGetter lightAccess, final ChunkAccess chunk, final ShortCollection sections) {
         final ChunkPos chunkPos = chunk.getPos();
-        final int chunkX = chunkPos.x();
-        final int chunkZ = chunkPos.z();
+        final int chunkX = chunkPos.x;
+        final int chunkZ = chunkPos.z;
 
         for (final ShortIterator iterator = sections.iterator(); iterator.hasNext();) {
             this.checkChunkEdge(lightAccess, chunk, chunkX, iterator.nextShort(), chunkZ);
@@ -595,8 +595,8 @@ public abstract class StarLightEngine {
     // This does not resolve skylight source problems.
     protected void checkChunkEdges(final LightChunkGetter lightAccess, final ChunkAccess chunk, final int fromSection, final int toSection) {
         final ChunkPos chunkPos = chunk.getPos();
-        final int chunkX = chunkPos.x();
-        final int chunkZ = chunkPos.z();
+        final int chunkX = chunkPos.x;
+        final int chunkZ = chunkPos.z;
 
         for (int currSectionY = toSection; currSectionY >= fromSection; --currSectionY) {
             this.checkChunkEdge(lightAccess, chunk, chunkX, currSectionY, chunkZ);
@@ -608,8 +608,8 @@ public abstract class StarLightEngine {
     // pulls light from neighbours, and adds them into the increase queue. does not actually propagate.
     protected final void propagateNeighbourLevels(final LightChunkGetter lightAccess, final ChunkAccess chunk, final int fromSection, final int toSection) {
         final ChunkPos chunkPos = chunk.getPos();
-        final int chunkX = chunkPos.x();
-        final int chunkZ = chunkPos.z();
+        final int chunkX = chunkPos.x;
+        final int chunkZ = chunkPos.z;
 
         for (int currSectionY = toSection; currSectionY >= fromSection; --currSectionY) {
             final SWMRNibbleArray currNibble = this.getNibbleFromCache(chunkX, currSectionY, chunkZ);
@@ -704,8 +704,8 @@ public abstract class StarLightEngine {
     }
 
     public final void forceHandleEmptySectionChanges(final LightChunkGetter lightAccess, final ChunkAccess chunk, final Boolean[] emptinessChanges) {
-        final int chunkX = chunk.getPos().x();
-        final int chunkZ = chunk.getPos().z();
+        final int chunkX = chunk.getPos().x;
+        final int chunkZ = chunk.getPos().z;
         this.setupCaches(lightAccess, chunkX * 16 + 7, 128, chunkZ * 16 + 7, true, true);
         try {
             // force current chunk into cache
@@ -754,8 +754,8 @@ public abstract class StarLightEngine {
     protected final boolean[] handleEmptySectionChanges(final LightChunkGetter lightAccess, final ChunkAccess chunk,
                                                         final Boolean[] emptinessChanges, final boolean unlit) {
         final Level world = (Level)lightAccess.getLevel();
-        final int chunkX = chunk.getPos().x();
-        final int chunkZ = chunk.getPos().z();
+        final int chunkX = chunk.getPos().x;
+        final int chunkZ = chunk.getPos().z;
 
         boolean[] chunkEmptinessMap = this.getEmptinessMap(chunkX, chunkZ);
         boolean[] ret = null;
@@ -905,8 +905,8 @@ public abstract class StarLightEngine {
     protected abstract void lightChunk(final LightChunkGetter lightAccess, final ChunkAccess chunk, final boolean needsEdgeChecks);
 
     public final void light(final LightChunkGetter lightAccess, final ChunkAccess chunk, final Boolean[] emptySections) {
-        final int chunkX = chunk.getPos().x();
-        final int chunkZ = chunk.getPos().z();
+        final int chunkX = chunk.getPos().x;
+        final int chunkZ = chunk.getPos().z;
         this.setupCaches(lightAccess, chunkX * 16 + 7, 128, chunkZ * 16 + 7, true, true);
 
         try {
@@ -955,8 +955,8 @@ public abstract class StarLightEngine {
         int lightCalls = 0;
 
         for (final ChunkPos chunkPos : chunks) {
-            final int chunkX = chunkPos.x();
-            final int chunkZ = chunkPos.z();
+            final int chunkX = chunkPos.x;
+            final int chunkZ = chunkPos.z;
             final ChunkAccess chunk = (ChunkAccess)lightAccess.getChunkForLighting(chunkX, chunkZ);
             if (chunk == null || !this.canUseChunk(chunk)) {
                 throw new IllegalStateException();
@@ -1198,7 +1198,7 @@ public abstract class StarLightEngine {
                             flags |= FLAG_HAS_SIDED_TRANSPARENT_BLOCKS;
                         }
 
-                        final int opacity = blockState.getLightDampening();
+                        final int opacity = blockState.getLightBlock();
                         final int targetLevel = propagatedLightLevel - Math.max(1, opacity);
                         if (targetLevel <= currentLevel) {
                             continue;
@@ -1280,7 +1280,7 @@ public abstract class StarLightEngine {
                             flags |= FLAG_HAS_SIDED_TRANSPARENT_BLOCKS;
                         }
 
-                        final int opacity = blockState.getLightDampening();
+                        final int opacity = blockState.getLightBlock();
                         final int targetLevel = propagatedLightLevel - Math.max(1, opacity);
                         if (targetLevel <= currentLevel) {
                             continue;
@@ -1407,7 +1407,7 @@ public abstract class StarLightEngine {
                             flags |= FLAG_HAS_SIDED_TRANSPARENT_BLOCKS;
                         }
 
-                        final int opacity = blockState.getLightDampening();
+                        final int opacity = blockState.getLightBlock();
                         final int targetLevel = Math.max(0, propagatedLightLevel - Math.max(1, opacity));
                         if (lightLevel > targetLevel) {
                             // it looks like another source propagated here, so re-propagate it
@@ -1536,7 +1536,7 @@ public abstract class StarLightEngine {
                             flags |= FLAG_HAS_SIDED_TRANSPARENT_BLOCKS;
                         }
 
-                        final int opacity = blockState.getLightDampening();
+                        final int opacity = blockState.getLightBlock();
                         final int targetLevel = Math.max(0, propagatedLightLevel - Math.max(1, opacity));
                         if (lightLevel > targetLevel) {
                             // it looks like another source propagated here, so re-propagate it
diff --git a/src/main/java/ca/spottedleaf/starlight/common/light/vanillainterface/BaseLevelLightEngineVanillaInterface.java b/src/main/java/ca/spottedleaf/starlight/common/light/vanillainterface/BaseLevelLightEngineVanillaInterface.java
index f821c0f..e12ab3e 100644
--- a/src/main/java/ca/spottedleaf/starlight/common/light/vanillainterface/BaseLevelLightEngineVanillaInterface.java
+++ b/src/main/java/ca/spottedleaf/starlight/common/light/vanillainterface/BaseLevelLightEngineVanillaInterface.java
@@ -22,7 +22,7 @@ import net.minecraft.world.level.chunk.LightChunkGetter;
 import net.minecraft.world.level.lighting.LayerLightEventListener;
 import net.minecraft.world.level.lighting.LayerLightSectionStorage;
 import net.minecraft.world.level.lighting.LevelLightEngine;
-import org.jspecify.annotations.Nullable;
+import org.jetbrains.annotations.Nullable;
 
 public class BaseLevelLightEngineVanillaInterface extends LevelLightEngine implements StarLightLightingProvider, ClientStarLightLightingProvider {
 
diff --git a/src/main/java/ca/spottedleaf/starlight/common/light/vanillainterface/CommonLightEngineUtils.java b/src/main/java/ca/spottedleaf/starlight/common/light/vanillainterface/CommonLightEngineUtils.java
index c403f31..e41ca1c 100644
--- a/src/main/java/ca/spottedleaf/starlight/common/light/vanillainterface/CommonLightEngineUtils.java
+++ b/src/main/java/ca/spottedleaf/starlight/common/light/vanillainterface/CommonLightEngineUtils.java
@@ -9,7 +9,7 @@ import net.minecraft.world.level.LightLayer;
 import net.minecraft.world.level.chunk.DataLayer;
 import net.minecraft.world.level.lighting.LayerLightEventListener;
 import net.minecraft.world.level.lighting.LayerLightSectionStorage;
-import org.jspecify.annotations.Nullable;
+import org.jetbrains.annotations.Nullable;
 
 public class CommonLightEngineUtils {
     private CommonLightEngineUtils() {
@@ -36,7 +36,7 @@ public class CommonLightEngineUtils {
     public static void setLightEnabled(StarLightLightingProvider instance, ChunkPos pos, boolean enable) {
         // store state for implementation of lightOnInColumn()
         // needed for proper culling of chunks in the client
-        final long key = pos.pack();
+        final long key = pos.toLong();
         if (enable) {
             instance.scalablelux$getLightingEnabledChunks().add(key);
         } else {
diff --git a/src/main/java/ca/spottedleaf/starlight/common/light/vanillainterface/ThreadedLevelLightEngineVanillaInterface.java b/src/main/java/ca/spottedleaf/starlight/common/light/vanillainterface/ThreadedLevelLightEngineVanillaInterface.java
index 1a99be8..1e4f8bf 100644
--- a/src/main/java/ca/spottedleaf/starlight/common/light/vanillainterface/ThreadedLevelLightEngineVanillaInterface.java
+++ b/src/main/java/ca/spottedleaf/starlight/common/light/vanillainterface/ThreadedLevelLightEngineVanillaInterface.java
@@ -25,7 +25,7 @@ import net.minecraft.world.level.chunk.LightChunkGetter;
 import net.minecraft.world.level.chunk.status.ChunkStatus;
 import net.minecraft.world.level.lighting.LayerLightEventListener;
 import net.minecraft.world.level.lighting.LayerLightSectionStorage;
-import org.jspecify.annotations.Nullable;
+import org.jetbrains.annotations.Nullable;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
@@ -294,7 +294,7 @@ public class ThreadedLevelLightEngineVanillaInterface extends ThreadedLevelLight
                 // can't really force the chunk to be edged checked, as we need neighbouring chunks - but we don't have
                 // them, so if it's not loaded then i guess we can't do edge checks. later loads of the chunk should
                 // catch what we miss here.
-                this.scalablelux$getLightEngine().checkChunkEdges(chunkPos.x(), chunkPos.z());
+                this.scalablelux$getLightEngine().checkChunkEdges(chunkPos.x, chunkPos.z);
             }
 
 //            this.chunkMap.releaseLightTicket(chunkPos); // vanilla 1.21 no longer does this
diff --git a/src/main/java/ca/spottedleaf/starlight/common/util/CoordinateUtils.java b/src/main/java/ca/spottedleaf/starlight/common/util/CoordinateUtils.java
index deebf1f..8100dd2 100644
--- a/src/main/java/ca/spottedleaf/starlight/common/util/CoordinateUtils.java
+++ b/src/main/java/ca/spottedleaf/starlight/common/util/CoordinateUtils.java
@@ -25,7 +25,7 @@ public final class CoordinateUtils {
     }
 
     public static long getChunkKey(final ChunkPos pos) {
-        return ((long) pos.z() << 32) | (pos.x() & 0xFFFFFFFFL);
+        return ((long) pos.z << 32) | (pos.x & 0xFFFFFFFFL);
     }
 
     public static long getChunkKey(final SectionPos pos) {
@@ -75,9 +75,9 @@ public final class CoordinateUtils {
     }
 
     public static long getChunkSectionKey(final ChunkPos pos, final int y) {
-        return ((pos.x() & SECTION_X_MASK) << SECTION_X_SHIFT)
+        return ((pos.x & SECTION_X_MASK) << SECTION_X_SHIFT)
                 | ((y & SECTION_Y_MASK) << SECTION_Y_SHIFT)
-                | ((pos.z() & SECTION_Z_MASK) << SECTION_Z_SHIFT);
+                | ((pos.z & SECTION_Z_MASK) << SECTION_Z_SHIFT);
     }
 
     public static long getChunkSectionKey(final BlockPos pos) {
diff --git a/src/main/java/ca/spottedleaf/starlight/mixin/client/multiplayer/ClientPacketListenerMixin.java b/src/main/java/ca/spottedleaf/starlight/mixin/client/multiplayer/ClientPacketListenerMixin.java
index b4621b5..016c403 100644
--- a/src/main/java/ca/spottedleaf/starlight/mixin/client/multiplayer/ClientPacketListenerMixin.java
+++ b/src/main/java/ca/spottedleaf/starlight/mixin/client/multiplayer/ClientPacketListenerMixin.java
@@ -110,7 +110,7 @@ public abstract class ClientPacketListenerMixin extends ClientCommonPacketListen
     )
     private void unloadLightDataHook(final ClientPacketListener instance, final ClientboundForgetLevelChunkPacket packet, final Operation<Void> original) {
         if (this.level.getChunkSource().getLightEngine() instanceof ClientStarLightLightingProvider clientStarLightLightingProvider) {
-            clientStarLightLightingProvider.scalablelux$clientRemoveLightData(new ChunkPos(packet.pos().x(), packet.pos().z()));
+            clientStarLightLightingProvider.scalablelux$clientRemoveLightData(new ChunkPos(packet.pos().x, packet.pos().z));
         } else {
             original.call(instance, packet);
         }
diff --git a/src/main/java/ca/spottedleaf/starlight/mixin/common/blockstate/BlockStateBaseMixin.java b/src/main/java/ca/spottedleaf/starlight/mixin/common/blockstate/BlockStateBaseMixin.java
index 3eaef26..9cd069c 100644
--- a/src/main/java/ca/spottedleaf/starlight/mixin/common/blockstate/BlockStateBaseMixin.java
+++ b/src/main/java/ca/spottedleaf/starlight/mixin/common/blockstate/BlockStateBaseMixin.java
@@ -1,6 +1,8 @@
 package ca.spottedleaf.starlight.mixin.common.blockstate;
 
 import ca.spottedleaf.starlight.common.blockstate.ExtendedAbstractBlockState;
+import com.mojang.serialization.MapCodec;
+import it.unimi.dsi.fastutil.objects.Reference2ObjectArrayMap;
 import net.minecraft.world.level.block.Block;
 import net.minecraft.world.level.block.state.BlockBehaviour;
 import net.minecraft.world.level.block.state.BlockState;
@@ -28,8 +30,8 @@ public abstract class BlockStateBaseMixin extends StateHolder<Block, BlockState>
     @Unique
     private boolean scalablelux$isConditionallyFullOpaque;
 
-    protected BlockStateBaseMixin(Block owner, Property<?>[] propertyKeys, Comparable<?>[] propertyValues) {
-        super(owner, propertyKeys, propertyValues);
+    protected BlockStateBaseMixin(Block object, Reference2ObjectArrayMap<Property<?>, Comparable<?>> reference2ObjectArrayMap, MapCodec<BlockState> mapCodec) {
+        super(object, reference2ObjectArrayMap, mapCodec);
     }
 
     /**
diff --git a/src/main/java/ca/spottedleaf/starlight/mixin/common/chunk/ChunkAccessMixin.java b/src/main/java/ca/spottedleaf/starlight/mixin/common/chunk/ChunkAccessMixin.java
index ac23fdc..66ad9be 100644
--- a/src/main/java/ca/spottedleaf/starlight/mixin/common/chunk/ChunkAccessMixin.java
+++ b/src/main/java/ca/spottedleaf/starlight/mixin/common/chunk/ChunkAccessMixin.java
@@ -5,7 +5,7 @@ import ca.spottedleaf.starlight.common.light.SWMRNibbleArray;
 import ca.spottedleaf.starlight.common.light.StarLightEngine;
 import ca.spottedleaf.starlight.common.light.StarLightLightingProvider;
 import com.llamalad7.mixinextras.injector.v2.WrapWithCondition;
-import net.minecraft.world.level.BlockAndLightGetter;
+import net.minecraft.world.level.BlockAndTintGetter;
 import net.minecraft.world.level.LevelAccessor;
 import net.minecraft.world.level.LevelHeightAccessor;
 import net.minecraft.world.level.chunk.ChunkAccess;
@@ -122,7 +122,7 @@ public abstract class ChunkAccessMixin implements ExtendedChunk {
         if (this.levelHeightAccessor instanceof LevelAccessor levelAccessor) {
             ChunkSource chunkSource = levelAccessor.getChunkSource();
             return chunkSource != null && chunkSource.getLightEngine() instanceof StarLightLightingProvider starLightLightingProvider;
-        } else if (this.levelHeightAccessor instanceof BlockAndLightGetter getter) {
+        } else if (this.levelHeightAccessor instanceof BlockAndTintGetter getter) {
             return getter.getLightEngine() instanceof StarLightLightingProvider starLightLightingProvider;
         } else {
             return false;
diff --git a/src/main/resources/fabric.mod.json b/src/main/resources/fabric.mod.json
index 3eb8b18..1613f98 100644
--- a/src/main/resources/fabric.mod.json
+++ b/src/main/resources/fabric.mod.json
@@ -32,6 +32,6 @@
   "accessWidener" : "scalablelux.accesswidener",
   "depends": {
     "fabricloader": ">=${loader_version}",
-    "minecraft": "~26.1-"
+    "minecraft": "=1.21.11"
   }
 }
diff --git a/src/main/resources/scalablelux.accesswidener b/src/main/resources/scalablelux.accesswidener
index b57b610..f676273 100644
--- a/src/main/resources/scalablelux.accesswidener
+++ b/src/main/resources/scalablelux.accesswidener
@@ -1,4 +1,4 @@
-accessWidener v1 official
+accessWidener v1 named
 
 # BlockStateBase.ShapeCache
 accessible class net/minecraft/world/level/block/state/BlockBehaviour$BlockStateBase$Cache
