From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: ishland <ishlandmc@yeah.net>
Date: Wed, 1 Jan 2025 12:36:45 +0800
Subject: [PATCH] fix: priority out of range


diff --git a/src/main/java/ca/spottedleaf/starlight/common/thread/SchedulingUtil.java b/src/main/java/ca/spottedleaf/starlight/common/thread/SchedulingUtil.java
index c9b3c58..cadd3c9 100644
--- a/src/main/java/ca/spottedleaf/starlight/common/thread/SchedulingUtil.java
+++ b/src/main/java/ca/spottedleaf/starlight/common/thread/SchedulingUtil.java
@@ -14,7 +14,7 @@ public class SchedulingUtil {
                 lockTokens.add(new LockTokenImpl(ownerTag, CoordinateUtils.getChunkKey(x + i, z + j)));
             }
         }
-        final SimpleTask simpleTask = new SimpleTask(task, lockTokens.toArray(LockToken[]::new), 240);
+        final SimpleTask simpleTask = new SimpleTask(task, lockTokens.toArray(LockToken[]::new), 60);
         GlobalExecutors.prioritizedScheduler.schedule(simpleTask);
     }
 
